assignments/burmanm/week-2.scm

```diff
@@ -0,0 +1,394 @@
+
+;; The time to calculate fermat numbers does not seem to increase at all and is less than the precision of
+;; the runtime. Replacing runtime with real-time-clock (mit-scheme) gives us a better indication:
+
+(define (timed-prime-test n)
+  (start-prime-test n (real-time-clock)))
+(define (start-prime-test n start-time)
+  (if (fast-prime? n 100)
+      (report-prime (- (real-time-clock) start-time) n)
+      #f))
+
+
+;; 10 error> (find-next-three (round->exact 1e40))
+
+;; 10000000000000000303786028427003666890753 *** 53
+;; 10000000000000000303786028427003666891041 *** 30
+;; 10000000000000000303786028427003666891101 *** 32
+;; ;Value: 3
+
+;; 10 error> (find-next-three (round->exact 1e300))
+
+;; 1000000000000000052504760255204420248704468581108159154915854115511802457988908195786371375080447864043704443832883878176942523235360430575644792184786706982848387200926575803737830233794788090059368953234970799945081119038967640880074652742780142494579258788820056842838115669472196386865459400540381 *** 502
+;; 1000000000000000052504760255204420248704468581108159154915854115511802457988908195786371375080447864043704443832883878176942523235360430575644792184786706982848387200926575803737830233794788090059368953234970799945081119038967640880074652742780142494579258788820056842838115669472196386865459400542127 *** 510
+;; 1000000000000000052504760255204420248704468581108159154915854115511802457988908195786371375080447864043704443832883878176942523235360430575644792184786706982848387200926575803737830233794788090059368953234970799945081119038967640880074652742780142494579258788820056842838115669472196386865459400543303 *** 496
+;; ;Value: 3
+
+;; 10 error> (find-next-three (round->exact 1e150))
+
+;; 999999999999999980835596172437374590573120014030318793091164810154100112203678582976298268616221151962702060266176005440567032331208403948233373515999 *** 151
+;; 999999999999999980835596172437374590573120014030318793091164810154100112203678582976298268616221151962702060266176005440567032331208403948233373516469 *** 151
+;; 999999999999999980835596172437374590573120014030318793091164810154100112203678582976298268616221151962702060266176005440567032331208403948233373517033 *** 150
+;; ;Value: 3
+
+;; Exercise 1.25
+
+;; Explained in the footnote 46, fast-exp would be a lot slower since we would have to calculate large numbers
+;; which is not necessary with the modulo operations.
+
+;; Exercise 1.26
+
+;; This transforms the expmod to be a tree recursion, which is a lot slower.
+
+;; Exercise 1.27
+
+(define (test-carmichael n)
+  (define (fermat-test n a)
+    (= (expmod a n n) a))
+  (define (iter-cm n a)
+    (if (= a n)
+	#t
+	(if (fermat-test n a)
+	    (iter-cm n (+ a 1))
+	    #f)))
+  (iter-cm n 2))
+
+(test-carmichael 561)
+
+;Value: #t
+
+(test-carmichael 1105)
+
+;Value: #t
+
+(test-carmichael 1729)
+
+;Value: #t
+
+(test-carmichael 2465)
+
+;Value: #t
+
+(test-carmichael 2821)
+
+;Value: #t
+
+(test-carmichael 6601)
+
+;Value: #t
+
```

Comment:
glad your method also evaluates true for 89. Not quite sure why this is the case :)

Reply:
89 is a prime?

Reply:
yes, but i thought the given procedure would only works for number that fools fermat-test.